src/transformers/cache_utils.py

```diff
@@ -0,0 +1,183 @@
+from typing import List, Optional, Tuple
+
+import torch
+
+
+class Cache:
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cos: Optional[torch.Tensor] = None,
+        sin: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        raise NotImplementedError("Make sure to implement `update` in a subclass.")
+
+
+class DynamicCache(Cache):
+    def __init__(self) -> None:
+        self.key_cache: List[Tuple[torch.Tensor]] = []
+        self.value_cache: List[Tuple[torch.Tensor]] = []
+
+    def __getitem__(self, key: int) -> List[Tuple[torch.Tensor]]:
+        """
+        Support for backwards-compatible `past_key_value` indexing, e.g. `past_key_value[0][0].shape[2]` to get the
+        sequence length.
+        """
+        if key == 0:
+            return self.key_cache
+        elif key == 1:
+            return self.value_cache
+        else:
+            raise KeyError(f"Cache only supports 0 (key) and 1 (value) indexing, got {key}")
+
+    def __iter__(self):
+        """
+        Support for backwards-compatible `past_key_value` iteration, e.g. `for x in past_key_value:` to iterate over
+        keys and values
+        """
+        yield self.key_cache
+        yield self.value_cache
+
+    def __len__(self):
+        """
+        Support for backwards-compatible `past_key_value` length, e.g. `len(past_key_value)`. This value corresponds
+        to the number of layers in the model.
+        """
+        return len(self.key_cache)
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cos: Optional[torch.Tensor] = None,
+        sin: Optional[torch.Tensor] = None,
```

Comment:
```suggestion
        pos_embeddings_dict: Optional[str, torch.Tensor]
```

Generally, I do agree that `cos`, `sin` is quite unique to RoPE embeddings. Another option would be to just add a general `pos_embeddings_dict: Optional[str, torch.Tensor]` to prevent the signature of `self.update` to explode.

Reply:
But passing dicts is also not ideal for `torch.compile` or `torch.trace.jit`